From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KCSesh <kssessio@amazon.com>
Date: Tue, 10 Mar 2026 04:00:00 +0000
Subject: [PATCH] fix const-correctness for C23 compatibility as implemented in
 glibc 2.43

glibc 2.43 C23 type-generic functions (memchr, bsearch, strrchr) now
return const-qualified pointers when given const-qualified input. Add
const qualifiers to remaining affected call sites in libdw and src.

[vighmah:
Reduced patch to only the two fixes not yet addressed by
4a5cf8be906d5991e7527e69e3f2ceaa74811301.
]
Signed-off-by: Vighnesh Maheshwari <vighmah@amazon.com>

---
 libdw/dwarf_getsrclines.c | 2 +-
 src/readelf.c             | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/libdw/dwarf_getsrclines.c b/libdw/dwarf_getsrclines.c
index 76db292..23942b8 100644
--- a/libdw/dwarf_getsrclines.c
+++ b/libdw/dwarf_getsrclines.c
@@ -527,7 +527,7 @@ read_srcfiles (Dwarf *dbg,
 
 	  /* First comes the file name.  */
 	  char *fname = (char *) linep;
-	  uint8_t *endp = memchr (fname, '\0', lineendp - linep);
+	  const uint8_t *endp = memchr (fname, '\0', lineendp - linep);
 	  if (endp == NULL)
 	    goto invalid_data;
 	  size_t fnamelen = endp - (uint8_t *) fname;
diff --git a/src/readelf.c b/src/readelf.c
index 0258d41..974a823 100644
--- a/src/readelf.c
+++ b/src/readelf.c
@@ -9674,7 +9674,7 @@ print_debug_line_section (Dwfl_Module *dwflmod, Ebl *ebl, GElf_Ehdr *ehdr,
 	    {
 	      /* First comes the file name.  */
 	      char *fname = (char *) linep;
-	      unsigned char *endp = memchr (fname, '\0', lineendp - linep);
+	      const unsigned char *endp = memchr (fname, '\0', lineendp - linep);
 	      if (unlikely (endp == NULL))
 		goto invalid_unit;
 	      linep = endp + 1;
